File: src/guardrails/normalize.js
// Text normalization — the FIRST step every guardrail runs before pattern
// matching. Attackers defeat naive regex filters by breaking up trigger words
// with characters that are invisible or look identical to ASCII:
//
//   • Zero-width joiners/spaces:   "ig<U+200B>nore previous instructions"
//   • Fullwidth / homoglyph forms: fullwidth "i" -> "i"
//   • Bidi overrides & BOM:        used to reorder or hide text
//   • Soft hyphen (U+00AD):        renders invisibly, splits words
//
// None of these survive Unicode NFKC folding + invisible-character stripping,
// so we canonicalize once here and let the detectors match plain ASCII. This is
// a recall win with essentially no false-positive cost: NFKC is a no-op on
// ordinary text, and legitimate content almost never carries zero-width control
// characters.

// Invisible / format code points that render to nothing but can split a trigger
// token. Built from numeric code points so this source file stays 100% ASCII
// (embedding the literal characters would make the file unreadable and easy to
// corrupt silently).
//   00AD soft hyphen · 200B-200F zero-width + bidi marks · 202A-202E bidi embeds
//   2060-2064 word-joiner/invisible math ops · 206A-206F deprecated format ·
//   FEFF BOM / zero-width no-break space
const INVISIBLE_SINGLES = [0x00ad, 0xfeff];
const INVISIBLE_RANGES = [
  [0x200b, 0x200f],
  [0x202a, 0x202e],
  [0x2060, 0x2064],
  [0x206a, 0x206f],
];

const hex = (c) => "\\u" + c.toString(16).padStart(4, "0");
const INVISIBLE = new RegExp(
  "[" +
    INVISIBLE_SINGLES.map(hex).join("") +
    INVISIBLE_RANGES.map(([a, b]) => `${hex(a)}-${hex(b)}`).join("") +
    "]",
  "g"
);

// Strip characters that render to nothing but break up patterns. Kept separate
// from full NFKC so callers that must preserve legitimate glyphs (e.g. model
// OUTPUT, where a "fi" ligature folding would be undesirable) can defang without
// folding.
export function stripInvisible(text = "") {
  const s = typeof text === "string" ? text : String(text ?? "");
  return s.replace(INVISIBLE, "");
}

// Canonical form: NFKC folds compatibility/fullwidth/homoglyph forms to their
// ASCII equivalents, then invisibles are removed. NFKC is idempotent, so
// re-normalizing already-normalized text is safe.
export function normalizeForAnalysis(text = "") {
  let s = typeof text === "string" ? text : String(text ?? "");
  try {
    s = s.normalize("NFKC");
  } catch {
    // Extremely old runtimes without String.prototype.normalize — fall back to
    // invisible-stripping alone rather than throwing inside a guardrail.
  }
  return stripInvisible(s);
}

File: backend/app/guardrails/normalize.py
"""Text normalization — the first step every detector runs.

Mirrors the frontend `normalize.js`: NFKC folding + invisible/zero-width
stripping, so homoglyph ("ｉgnore") and zero-width ("ig<ZWSP>nore") evasions fold
to plain ASCII before any pattern matches. Kept in sync with the client so the
optimistic UX pre-check and the authoritative server check agree.
"""

from __future__ import annotations

import re
import unicodedata

# Invisible / format code points that render to nothing but split trigger tokens.
# Built from code points so this source file stays pure ASCII.
#   00AD soft hyphen · 200B-200F zero-width + bidi · 202A-202E bidi embeds
#   2060-2064 word-joiner/invisible ops · 206A-206F deprecated format · FEFF BOM
_INVISIBLE_SINGLES = [0x00AD, 0xFEFF]
_INVISIBLE_RANGES = [(0x200B, 0x200F), (0x202A, 0x202E), (0x2060, 0x2064), (0x206A, 0x206F)]
_INVISIBLE = re.compile(
    "["
    + "".join(chr(c) for c in _INVISIBLE_SINGLES)
    + "".join(f"{chr(a)}-{chr(b)}" for a, b in _INVISIBLE_RANGES)
    + "]"
)


def strip_invisible(text: str) -> str:
    """Remove zero-width / invisible characters (preserves visible glyphs)."""
    return _INVISIBLE.sub("", text or "")


def normalize_for_analysis(text: str) -> str:
    """Canonical form for INPUT analysis: NFKC fold, then strip invisibles.

    NFKC is idempotent, so repeated normalization down the pipeline is safe.
    """
    if not text:
        return ""
    return strip_invisible(unicodedata.normalize("NFKC", text))
